File: 1331-rank-transform-of-an-array/1331-rank-transform-of-an-array.cpp
class Solution {
public:
    vector<int> arrayRankTransform(vector<int>& arr) {
        int n = arr.size();
        map<int, vector<int>> numToIndex;
        for(int i = 0; i < n; i++){
            numToIndex[arr[i]].push_back(i);
        }

        int rank = 1;
        vector<int> ans(n, 0);
        for(auto it: numToIndex){
            for(int index: it.second){
                ans[index] = rank;
            }
            rank++;
        }

        return ans;
    }
};

File: 1331-rank-transform-of-an-array/1331-rank-transform-of-an-array.py
class Solution:
    def arrayRankTransform(self, arr: List[int]) -> List[int]:
        dic = defaultdict(list)
        for i in range(len(arr)):
            dic[arr[i]].append(i)
        
        dic = dict(sorted(dic.items()))

        rank = 1
        ans = [0] * len(arr)
        for element, indices in dic.items():
            for index in indices:
                ans[index] = rank
            rank += 1
        return ans
        